File: 3/solution.js
var init = require('../shared/init');

init('input.txt', [
  countHousesVisited,
  { fn: countHousesVisited, args: [2] }
]);

function countHousesVisited(input, numSantas) {
  var numSantas = numSantas || 1;
  var santaTracker = buildSantaTracker(numSantas);

  input.split('').forEach(function(char, i) {
    var santa = santaTracker[i % numSantas];
    var currentLoc = santa.currentLoc;
    currentLoc = directionMap[char](currentLoc);

    santa.visited.add(currentLoc.join(','));
  });

  return zipSantas(santaTracker).size;
}

// helpers

function zipSantas(santas) {
  var keys = Object.keys(santas);

  if (keys.length === 1) {
    return santas["0"].visited;
  }

  return new Set(
    keys.reduce(function(arr, id) {
      var visited = santas[id].visited;

      return arr.concat([...visited]);
    }, [])
  );
}

function buildSantaTracker(n) {
  var santaTracker = {};

  for (var i = 0; i < n; i++) {
    santaTracker[i] = {
      currentLoc: [0, 0],
      visited: new Set()
    };
  }

  return santaTracker;
}

var directionMap = {
  '^': function(loc) { return [++loc[0], loc[1]]; },
  'v': function(loc) { return [--loc[0], loc[1]]; },
  '>': function(loc) { return [loc[0], ++loc[1]]; },
  '<': function(loc) { return [loc[0], --loc[1]]; }
};

File: 3/solution.rb
require "../shared/read_file.rb"
require "set"

input = read_file("input.txt")

def count_houses_visited(input, num_santas = 1)
  visited = Set.new
  santa_tracker = build_santa_tracker(num_santas)

  input.each_char.with_index do |char, i|
    loc = santa_tracker[i % num_santas];

    if char === "^"
      loc[0] += 1
    elsif char === "v"
      loc[0] -= 1
    elsif char === ">"
      loc[1] += 1
    elsif char === "<"
      loc[1] -= 1
    end

    visited.add(loc.join(","))
  end

  visited.size
end

# helpers

def build_santa_tracker(n)
  santa_tracker = Hash.new

  n.times do |n|
    santa_tracker[n] = [0, 0]
  end

  santa_tracker
end

puts "Part 1: #{count_houses_visited(input)}"
puts "Part 2: #{count_houses_visited(input, 2)}"
